Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.cpp

```diff
@@ -84,106 +85,195 @@ void NodeWiseMeshPartitioner::readMetisData(const std::string& file_name_base)
+    {
+        const auto* elem = elements[elem_id];
+        if (_elements_status[elem_id])
+            continue;
 
-        // Find non-ghost nodes in this partition
-        // -- Extra nodes for high order elements
-        std::vector<MeshLib::Node*> extra_nodes;
-        for (std::size_t i = 0; i < _mesh->getNumberOfNodes(); i++)
+        std::size_t non_ghost_node_number = 0;
+        for (unsigned i = 0; i < elem->getNumberOfNodes(); i++)
         {
-            if (_nodes_partition_ids[i] == part_id)
+            if (_nodes_partition_ids[elem->getNodeIndex(i)] == part_id)
             {
-                if (is_mixed_high_order_linear_elems)
-                { // TODO: Test it once there is a case
-                    if (i < _mesh->getNumberOfBaseNodes())
-                        partition.nodes.push_back(nodes[i]);
-                    else
-                        extra_nodes.push_back(nodes[i]);
-                }
-                else
-                {
-                    partition.nodes.push_back(nodes[i]);
-                }
+                non_ghost_node_number++;
             }
         }
-        partition.number_of_non_ghost_base_nodes = partition.nodes.size();
-        partition.number_of_non_ghost_nodes =
-            partition.number_of_non_ghost_base_nodes + extra_nodes.size();
 
-        // Find elements that belong to this partition
-        std::vector<MeshLib::Element*> const& elements = _mesh->getElements();
-        for (std::size_t elem_id = 0; elem_id < elements.size(); elem_id++)
+        if (non_ghost_node_number == 0)
+            continue;
+
+        if (non_ghost_node_number == elem->getNumberOfNodes())
         {
-            const auto* elem = elements[elem_id];
-            if (_elements_status[elem_id])
+            partition.regular_elements.push_back(elem);
+            _elements_status[elem_id] = true;
+        }
+        else
+        {
+            partition.ghost_elements.push_back(elem);
+        }
+    }
+}
+
+void NodeWiseMeshPartitioner::findGhostNodesInPartition(
+    std::size_t part_id,
+    const bool is_mixed_high_order_linear_elems,
+    std::vector<MeshLib::Node*>& extra_nodes)
+{
+    auto& partition = _partitions[part_id];
+    std::vector<MeshLib::Node*> const& nodes = _mesh->getNodes();
+    std::vector<bool> nodes_reserved(_mesh->getNumberOfNodes(), false);
+    for (const auto* ghost_elem : partition.ghost_elements)
+    {
+        for (unsigned i = 0; i < ghost_elem->getNumberOfNodes(); i++)
+        {
+            const unsigned node_id = ghost_elem->getNodeIndex(i);
+            if (nodes_reserved[node_id])
                 continue;
 
-            std::size_t non_ghost_node_number = 0;
-            for (unsigned i = 0; i < elem->getNumberOfNodes(); i++)
+            if (_nodes_partition_ids[node_id] != part_id)
             {
-                if (_nodes_partition_ids[elem->getNodeIndex(i)] == part_id)
+                if (is_mixed_high_order_linear_elems)
```

Comment:
This is repeated (or very similar to above). Is it possible to extract a method?